docs/Foundation/NSInvocation.md

```diff
@@ -0,0 +1,127 @@
+# Windows Bridge for iOS Dev Design Specification: NSInvocation
+
+## Feature Summary
+
+**NSInvocation** is a Foundation class that provides an object representation of a method call.
+It encapsulates a target, a selector, and a pack of arguments whose contents can be installed
+or interrogated at runtime.
+
+## Challenges
+
+Objective-C methods are implemented as though they were C functions, with standard C function calling conventions, taking 2+n arguments.
+The first two arguments passed to a method implementation are always `self` (the target of the message) and `_cmd` (the message selector).
+The remaining arguments are laid out according to the platform's default calling convention -- `__cdecl` on Win32 and *AAPCS*<sup>1</sup> on ARM.
+
+Methods that return structures above a certain *platform-defined* length limit are emitted with a hidden first argument, shifting `self` and `_cmd`
+one machine word up to make room for a *structure return* pointer. This out pointer is populated with the return value of the method automatically.
+
+Herein lies the challenge: **NSInvocation** must be able to lay out arguments at runtime, based solely on the Objective-C method signature
+provided during its construction.
+
+<sup>1. Procedure Call Standard for the ARM Architecture. [PDF link](http://infocenter.arm.com/help/topic/com.arm.doc.ihi0042f/IHI0042F_aapcs.pdf)</sup>
+
+## Architecture Decisions
+
+NSInvocation will rely on an internal platform-specific interface, `_NSInvocationCallFrame`, to do its argument allocation.
+
+`_NSInvocationCallFrame` has only three responsibilities: determining whether a method requires special struct return handling, storing an argument, and loading an argument.
+
+Therefore, platform support for NSInvocation will be provided via two discrete modules:
+
+1. A conformant implementation of `_NSInvocationCallFrame`.
+2. An assembly thunk that will take the platform internal representation of `_NSInvocationCallFrame` and dispatch a call to a function pointer with the stored arguments.
+
+### X86 details
+
+X86 is the easier of the two currently-supported architectures for which to allocate arguments.
+
+* Every argument is placed on the stack, in stack-ascending order.
+* Arguments less than four bytes in width are sign- or zero-extended and stored in one doubleword and emitted onto the stack.
```

Comment:
>doubleword [](start = 87, length = 10)

doubleword == 32 bits for x86? it may be better to call out the actual width here to avoid any confusions with Win32 DWORD.